chore(db, model): simplify per hash DB lookup in copier

imsodin opened the pull request:
This is a draft because I haven't adjusted all the tests yet, I'd like to get feedback on the change overall first, before spending time on that.

In my opinion the main win of this change is in it's lower complexity
resp. fewer moving parts. It should also be faster as it only does one
query instead of two, but I have no idea if that's practically
relevant.

This also mirrors the v1 DB, where a block map key had the name
appended. Not that this is an argument for the change, it was mostly
reassuring me that I might not be missing something key here
conceptually (I might still be of course, please tell me :) ).

And the change isn't mainly intrinsically motivated, instead it came
up while fixing a bug in the copier. And the nested nature of that code
makes the fix harder, and "un-nesting" it required me to understand
what's happening. This change fell out of that.

calmh commented:
Looks very reasonable

imsodin commented:
Thanks for confirming, let me do the annoying part then :)

imsodin committed:
chore(db, model): Simplify per hash DB lookup in copier

In my opinion the main win of this change is in it's lower complexity
resp. fewer moving parts. It should also be faster as it only does one
query instead of two, but I have no idea if that's practically
relevant.

This also mirrors the v1 DB, where a block map key had the name
appended. Not that this is an argument for the change, it was mostly
reassuring me that I might not be missing something key here
conceptually.

This came up while working on another bug in the copier (open file
leak). And the nested nature of this code makes the fix harder, and
"un-nesting" it required me to understand what's happening. This change
fell out of that.

imsodin committed:
Refactor to iter based signature for consistency

imsodin commented:
I might have done a bit more than just adjust the errors: I noticed this blocks function being the only one that doesn't use an `iter` based signature. And switching to that also eliminated one nested loop, and thus named also named loop, in the copier code. Admittedly I could also get the same effect without `iter` by flattening the returned map. If there's a reason not to use `iter` here, please let me know.

calmh reviewed (approved):
Sweet

imsodin commented:
Yeah. And in the "normal" case where the first block can be copied, `iter` even avoids db queries if the block also exists in other folders.

calmh commented:
Looks like we get a nil folderFilesystem, which I was actually wondering if that could happen, but didn't look like something specifically introduced in your change...

imsodin commented:
Well, it can happen if one forgets to set the folder ID :P  
Which now makes me doubt if that flat structure is really such a nice decision for the DB interface. In any case, I'll fix it somehow, soon.

imsodin committed:
straightforward fix

imsodin committed:
More consistency and simplicity: everything by folder

If I do really care about the nesting in the copier, I can get rid of
that independently.

calmh reviewed (approved):
Yup
